# Задание 5

In [ ]:
import numpy as np
import random


N = 100
np.random.seed(42)
tetha = np.random.randint(100)  # сюда тету вставляете
print(f"Θ = {tetha}")
X = np.random.randint(tetha, 2 * tetha, N)

tetha_s_volnoi = 2 / 3 * (np.sum(X) / N)
a = 0.05

l0 = np.max(X) / (1 + (1 - a / 2) ** (1 / N))
r0 = np.max(X) / (1 + (a / 2) ** (1 / N))
print(f"Точный доверительный интервал: ({l0:.4f}, {r0:.4f}; размах {r0 - l0:.4f}")

l1 = -1.96 * (2 / 3) / 10 * np.sqrt(0.01 * np.sum(X ** 2) - (0.01 * np.sum(X)) ** 2) + tetha_s_volnoi
r1 = 1.96 * (2 / 3) / 10 * np.sqrt(0.01 * np.sum(X ** 2) - (0.01 * np.sum(X)) ** 2) + tetha_s_volnoi
print(f"Асимптотический доверительный интервал: ({l1:.4f}, {r1:.4f}); pазмах {r1 - l1:.4f}")

S = 100000       # миллион оно считало 34 минуты, мой ноут начал шипеть где-то на 10, больше 100к больше не генерирую
M = np.array([])
for i in range(S):
    M = np.append(M, 2 / 3 * np.mean(np.random.choice(X, size=len(X))) - 2 / 3 * np.mean(X))
M = np.sort(M)

l2 = 2 / 3 * np.mean(X) + M[24]
r2 = 2 / 3 * np.mean(X) + M[974]

print(f"Бутстраповский недоверительный интервал: [{l2:.4f}, {r2:.4f}]; pазмах {r2 - l2:.4f}")



Θ = 51
Точный доверительный интервал: (50.5064, 51.4313; размах 0.9249
Асимптотический доверительный интервал: (48.5060, 52.3340); pазмах 3.8281
Бутстраповский недоверительный интервал: [46.9733, 48.1467]; pазмах 1.1733


Точный доверительный интервал: (50.5064, 51.4313); размах 0.9249

Асимптотический доверительный интервал: (48.5060, 52.3340); pазмах 3.8281

Бутстраповский недоверительный интервал: [46.5667, 47.4133]; pазмах 0.8467   на S = миллион

### Сравнил, при 10000 бутстрапов он похож на ассимптотический0, но при 100000 он уже съезжает куда-то не туда и начинает быть узким огогошеньки, на миллоне так ваще хана. Но в общем-то ассимптотический по крайней мере покрывает точный интервал, но имеет максимальную ширину. Точный - узкий и хорош и вообще лапочка. Бустрап че-то среднее но иногда не туда. На миллионом бутстрапе размах оказался меньше, чем в точном, что настораживает

# Задание 6

In [18]:
import numpy as np

def f1(x, tetha):
    return (1 - x) ** (1 / (1 - tetha))
def f2(X, N):
    return 1 + N / sum([np.log(x) for x in X])


np.random.seed(42)
tetha = 3  # сюда тету вставляете
N = 100
print(f"θ = {tetha}")
a = 0.05
X.sort()

tetha_s_volnoi = N / np.sum(np.log(X)) + 1

l1 = (-(1.96 * np.log(2) * (2 ** (1 / (tetha_s_volnoi - 1)))) / (tetha_s_volnoi - 1) / 10 + 2 ** (1 / (tetha_s_volnoi - 1)))
r1 = (1.96 * np.log(2) * (2 ** (1 / (tetha_s_volnoi - 1)))) / (tetha_s_volnoi - 1) / 10 + 2 ** (1 / (tetha_s_volnoi - 1))
print(f"Доверительный интервал для медианы: [{l1:.4f}, {r1:.4f}]; размах {r1 - l1:.4f}")


l2 = -1.96 * (tetha_s_volnoi - 1) / np.sqrt(N) + tetha_s_volnoi
r2 = 1.96 * (tetha_s_volnoi - 1) / np.sqrt(N) + tetha_s_volnoi
print(f"Ассимптотический доверительный интервал: [{l2:.4f}, {r2:.4f}]; размах {r2 - l2:.4f}")

M = np.sort(np.array([f2([f1(x, tetha) for x in np.random.sample(N)], N) for i in range(50000)]))

l3 = M[int(a / 2 * 50000)]
r3 = M[int((1 - a / 2) * 50000)]
print(f"Бутстраповский-параметрический доверительныйинтервал: [{l3:.4f},  {r3:.4f}]; размах {r3 - l3:.4f}")

M = np.array([])
S = 100000      # Миллион даже не пытался сгенерировать
for i in range(S):
    M = np.append(M, N / np.sum(np.log((np.random.choice(X, size = len(X))))) + 1 - tetha_s_volnoi)
M = np.sort(M)

l4 = tetha_s_volnoi + M[74]
r4 = tetha_s_volnoi + M[974]
print(f"Бутстраповский-непараметрический доверительный интервал: [{l4:.4f}, {r4:.4f}; размах {r4 - l4:.4f}")

θ = 3
Доверительный интервал для медианы: [1.2877, 1.4584]; размах 0.1706
Ассимптотический доверительный интервал: [2.7579, 3.6149]; размах 0.8571
Бутстраповский-параметрический доверительныйинтервал: [2.6591,  3.4572]; размах 0.7981
Бутстраповский-непараметрический доверительный интервал: [2.6333, 2.7556; размах 0.1223
